tools/src/org.graalvm.tools.api.lsp/src/org/graalvm/tools/api/lsp/LSPCommand.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.tools.api.lsp;
+
+import java.util.List;
+
+import com.oracle.truffle.api.instrumentation.TruffleInstrument.Env;
+
+/**
+ * Command interface provided by a {@link LSPExtension}.
+ */
+public interface LSPCommand {
+
+    /**
+     * Get the name of the command.
+     */
+    String getName();
+
+    /**
+     * Execute the command.
+     */
+    Object execute(LSPServerAccessor server, Env env, List<Object> arguments);
```

Comment:
With `Env` this is quite powerful. This allows an arbitrary instrumentation and can potentially be abused. How it's used? Would some restricted access be sufficient?

Reply:
The `Env` object passed in is the `envInternal` and I exposed it because it *is* quite powerful. The idea of `LSPExtension`s is that they can parse, evaluate, and instrument code within the context of the GraalLS. Of course, we could hide some of `Env`'s functionality similar to how the `LSPServerAccessor` hides information. For our tools, we currently only need `#getInstrumenter()`, `#parse()`, and `#parseInline()`.

BTW: I wonder, can't other instruments just create `Context` objects through the Graal SDK and do arbitrary instrumentation? I mean all instruments must be trusted, right?